HW2/P4/driver.py

```diff
@@ -0,0 +1,89 @@
+import sys
+import os.path
+sys.path.append(os.path.join('..', 'util'))
+
+import set_compiler
+set_compiler.install()
+
+import pyximport
+pyximport.install()
+
+import numpy as np
+import pylab
+
+import filtering
+from timer import Timer
+import threading
+
+def py_median_3x3(image, iterations=10, num_threads=1):
+    # repeatedly filter with a 3x3 median
+    tmpA = image.copy()
+    tmpB = np.empty_like(tmpA)
+
+    # Each thread controls every Nth line in image
+    threads = [None] * num_threads
+    # Each event corresponds to a particular thread and particular iteration - represents whether that thread
+    # is done with that iteration
+    events = [[threading.Event() for i in range(iterations)] for j in range(num_threads)]
+
+    for thread_id in range(num_threads):
+        threads[thread_id] = threading.Thread(target=filter_image, args=(num_threads, thread_id, tmpA, tmpB, events, iterations))
+        threads[thread_id].start()
+
+    # Terminate all threads
+    for thread in threads:
+        thread.join()
+
+    # Return fully filtered image
+    return tmpA
+
+def filter_image(num_threads, thread_id, tmpA, tmpB, events, iterations=10):
+    for itr in range(iterations):
+        # Waits on itr-1 of thread_id-1 and thread_id+1 to finish itr-1 before executing current itr
+        if num_threads > 1 and itr > 0:
```

Comment:
Nice conditioning to reduce the code for the function which you pass to the threads. I guess I wrote a condition too many.